matrix_sdk_appservice/src/warp.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2021 Famedly GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{net::ToSocketAddrs, result::Result as StdResult};
+
+use futures::TryFutureExt;
+use matrix_sdk::Bytes;
+use serde::Serialize;
+use warp::{filters::BoxedFilter, path::FullPath, Filter, Rejection, Reply};
+
+use crate::{Appservice, Error, Result};
+
+pub async fn run_server(
+    appservice: Appservice,
+    host: impl Into<String>,
+    port: impl Into<u16>,
+) -> Result<()> {
+    let routes = warp_filter(appservice);
+
+    let mut addr = format!("{}:{}", host.into(), port.into()).to_socket_addrs()?;
+    if let Some(addr) = addr.next() {
+        warp::serve(routes).run(addr).await;
+        Ok(())
+    } else {
+        Err(Error::HostPortToSocketAddrs)
+    }
+}
+
+pub fn warp_filter(appservice: Appservice) -> BoxedFilter<(impl Reply,)> {
+    warp::any()
+        .and(filters::valid_access_token(appservice.registration().hs_token.clone()))
+        .and(filters::transactions(appservice))
+        .or(filters::users())
+        .or(filters::rooms())
+        .recover(handle_rejection)
+        .boxed()
+}
+
+mod filters {
+    use super::*;
+
+    pub fn users() -> BoxedFilter<(impl Reply,)> {
+        warp::get()
+            .and(
+                warp::path!("_matrix" / "app" / "v1" / "users" / String)
+                    // legacy route
+                    .or(warp::path!("users" / String))
```

Comment:
`.and(warp::path::end())` is missing here and for the other `path!` matchers

Reply:
The `path!` macro actually already does `path::end()` internally, but doesn't hurt to have it twice in case a non-macro path matcher gets added